# Assignment 1
You should submit the **UniversityNumber.ipynb** file and your final prediction file **UniversityNumber.test.txt** to moodle. Make sure your code does not use your local files and that the results are reproducible. Before submitting, please **1. clean all outputs and 2. run all cells in your notebook and keep all running logs** so that we can check.

# 1 $n$-gram Language Model

In [1]:
!mkdir -p data/lm
!wget -O data/lm/train.txt https://raw.githubusercontent.com/ranpox/comp3361-spring2024/main/assignments/A1/data/lm/train.txt
!wget -O data/lm/dev.txt https://raw.githubusercontent.com/ranpox/comp3361-spring2024/main/assignments/A1/data/lm/dev.txt
!wget -O data/lm/test.txt https://raw.githubusercontent.com/ranpox/comp3361-spring2024/main/assignments/A1/data/lm/test.txt

The syntax of the command is incorrect.
'wget' is not recognized as an internal or external command,
operable program or batch file.
'wget' is not recognized as an internal or external command,
operable program or batch file.
'wget' is not recognized as an internal or external command,
operable program or batch file.


## 1.1 Building vocabulary

### Code

In [23]:
with open('data/lm/train.txt', 'r', encoding="utf-8") as f:
    train = dict()
    for line in f:
        line = '<START> ' + line + ' <END>'     # to add <START> and <END> to the start and end of each sentence 
        tmp = line.split()
        for word in tmp:
            if word in train.keys():            # if in dict, increment counter
                train[word] += 1
            else:
                train[word] = 1                 # if not in dict, add new entry

    countUNK = 0
    for key in list(train.keys()):              # find all words with freq < 3, sum their counts
        if train[key] < 3:
            countUNK += train[key]
            del train[key]
    train['<UNK>'] = countUNK
    print("Size of vocabulary = " + str(len(train)))

Size of vocabulary = 26601


### Discussion

Size of vocabulary = 26601

Number of parameters of n-gram models = (size of vocabulary) ^ n = (26601) ^ n

## 1.2 $n$-gram Language Modeling

### Code

### Discussion

## 1.3 Smoothing

### 1.3.1 Add-one (Laplace) smoothing

### Code

#### Discussion

#### 1.3.2 Add-k smoothing

##### Code

##### Discussion

### 1.3.3 Linear Interpolation

#### Code

#### Discussion

##### 1.3.4 Optimization

#### Discussion

# 2 Word Vectors

In [3]:
def load_embedding_model():
    """ Load GloVe Vectors
        Return:
            wv_from_bin: All 400000 embeddings, each lengh 50
    """
    import gensim.downloader as api
    wv_from_bin = api.load("glove-wiki-gigaword-200")
    print("Loaded vocab size %i" % len(list(wv_from_bin.index_to_key)))
    return wv_from_bin

wv_from_bin = load_embedding_model()

ModuleNotFoundError: No module named 'gensim'

## 2.1 Find most similar word
Use cosine similarity to find the most similar word to each of these words. Report the most similar word and its cosine similarity.

## 2.2 Finding Analogies
Use vector addition and subtraction to compute target vectors for the analogies below. After computing each target vector, find the top three candidates by cosine similarity. Report the candidates and their similarities to the target vector.

- dog : puppy :: cat : ?:
- speak : speaker :: sing : ?:
- France : French :: England : ?:
- France : wine :: England : ?

# 3 Sentiment analysis

In [ ]:
!mkdir -p data/classification
!wget -O data/classification/train.txt https://raw.githubusercontent.com/ranpox/comp3361-spring2024/main/assignments/A1/data/classification/train.txt
!wget -O data/classification/dev.txt https://raw.githubusercontent.com/ranpox/comp3361-spring2024/main/assignments/A1/data/classification/dev.txt
!wget -O data/classification/test-blind.txt https://raw.githubusercontent.com/ranpox/comp3361-spring2024/main/assignments/A1/data/classification/test-blind.txt

## 3.1 Logistic Regression

### Unigram Features

### Bigram Features

### GloVe Features

Compare the performance of three types of features on dev set. Report the weighted average precision, recall and F1-score for each feature set.

| Feature | precision | recall | F1-score |
| ----------- | --------- | ------ | -------- |
| unigram     |           |        |          |
| bigram      |           |        |          |
| GloVe       |           |        |          |

## 3.2 Better Feature